engine/routines.go

```diff
@@ -252,121 +251,124 @@ func WebsocketRoutine() {
-					Bot.ExchangeCurrencyPairManager.update(ws.GetName(),
-						result.Pair,
-						result.Asset,
-						SyncItemOrderbook,
-						nil)
-				}
+			err := WebsocketDataHandler(ws.GetName(), data)
+			if err != nil {
+				log.Error(log.WebsocketMgr, err)
+			}
+		}
+	}
+}
 
-				if Bot.Settings.Verbose {
-					log.Infof(log.WebsocketMgr,
-						"%s websocket %s %s orderbook updated\n",
-						ws.GetName(),
-						FormatCurrency(result.Pair),
-						d.Asset)
-				}
-			default:
-				if Bot.Settings.Verbose {
-					log.Warnf(log.WebsocketMgr,
-						"%s websocket Unknown type: %+v\n",
-						ws.GetName(),
-						d)
-				}
+func WebsocketDataHandler(exchName string, data interface{}) error {
+	if data == nil {
+		return fmt.Errorf("routines.go - exchange %s nil data sent to websocket",
+			exchName)
+	}
+	switch d := data.(type) {
+	case string:
+		switch d {
+		case wshandler.WebsocketNotEnabled:
+			return fmt.Errorf("routines.go - exchange %s websocket not enabled",
+				exchName)
+		default:
+			log.Info(log.WebsocketMgr, d)
+		}
+	case error:
+		return fmt.Errorf("routines.go exchange %s websocket error - %s", exchName, data)
+	case wshandler.TradeData:
+		if Bot.Settings.Verbose {
+			log.Infof(log.WebsocketMgr, "%s websocket %s %s trade updated %+v",
+				exchName,
+				FormatCurrency(d.CurrencyPair),
+				d.AssetType,
+				d)
+		}
+	case wshandler.FundingData:
+		if Bot.Settings.Verbose {
+			log.Infof(log.WebsocketMgr, "%s websocket %s %s funding updated %+v",
+				exchName,
+				FormatCurrency(d.CurrencyPair),
+				d.AssetType,
+				d)
+		}
+	case *ticker.Price:
+		if Bot.Settings.EnableExchangeSyncManager && Bot.ExchangeCurrencyPairManager != nil {
+			Bot.ExchangeCurrencyPairManager.update(exchName,
+				d.Pair,
+				d.AssetType,
+				SyncItemTicker,
+				nil)
+		}
+		err := ticker.ProcessTicker(exchName, d, d.AssetType)
+		printTickerSummary(d, d.Pair, d.AssetType, exchName, "websocket", err)
+	case wshandler.KlineData:
+		if Bot.Settings.Verbose {
+			log.Infof(log.WebsocketMgr, "%s websocket %s %s kline updated %+v",
+				exchName,
+				FormatCurrency(d.Pair),
+				d.AssetType,
+				d)
+		}
+	case wshandler.WebsocketOrderbookUpdate:
+		if Bot.Settings.EnableExchangeSyncManager && Bot.ExchangeCurrencyPairManager != nil {
+			Bot.ExchangeCurrencyPairManager.update(exchName,
+				d.Pair,
+				d.Asset,
+				SyncItemOrderbook,
+				nil)
+		}
+
+		if Bot.Settings.Verbose {
+			log.Infof(log.WebsocketMgr,
+				"%s websocket %s %s orderbook updated",
+				exchName,
+				FormatCurrency(d.Pair),
+				d.Asset)
+		}
+	case *order.Detail:
+		if !Bot.OrderManager.orderStore.existsWithLock(d) {
+			err := Bot.OrderManager.orderStore.Add(d, true)
+			if err != nil {
+				return err
 			}
+		} else {
+			od, err := Bot.OrderManager.orderStore.GetByExchangeAndID(d.Exchange, d.ID)
+			if err != nil {
+				return err
+			}
+			od.UpdateOrderFromDetail(d)
+		}
+	case *order.Cancel:
+		err := Bot.OrderManager.Cancel(d)
```

Comment:
as nothing further is being done here can just return 

This can be done for a few of the returns in this function if you prefere return early style since after the switch return is always nil